### PR TITLE
chore!: migrate escrow accounts

### LINKED ISSUE #2989: Migrate escrow accounts with discrepancies
### Problem

A recent issue found in the PFM issue created some discrepancies between the transfer module's escrow accounts and the counterparty chains' supply. 

### Closing criteria

Mint and transfer the missing assets to the affected escrow accounts.

### PR DESCRIPTION
**Pull request by sainoe**
## Description

Closes: #2989

cc @jtieri

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] Added `!` to the type prefix if API, client, or state breaking change (i.e., requires minor or major version bump)
* [ ] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] Provided a link to the relevant issue or specification
* [ ] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
* [ ] Included the necessary unit and integration [tests](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry in `.changelog` (for details, see [contributing guidelines](../../CONTRIBUTING.md#changelog))
* [ ] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [ ] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed all author checklist items have been addressed 
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

File: app/upgrades/v15/upgrades.go [hunk 1]
```diff
@@ -350,3 +356,94 @@ func SetMinInitialDepositRatio(ctx sdk.Context, gk govkeeper.Keeper) error {
...
+- Escrow Account Address: cosmos1x54ltnyg88k0ejmk8ytwrhd3ltm84xehrnlslf
+- Asset Base Denom: FX
+- Asset IBC Denom: ibc/4925E6ABA571A44D2BE0286D2D29AF42A294D0FF2BB16490149A1B26EAD33729
+- Escrow Balance: 8859960534331100342
+- Counterparty Total Supply: 8899960534331100342ibc/EBBE6553941A1F0111A9163F885F7665417467FB630D68F5D4F15425C1E64FDE
+- Missing amount in Escrow Account: 40000000000000000
+
+Discrepancy #2:
+- Counterparty Chain ID: osmosis-1
+- Escrow Account Address: cosmos1x54ltnyg88k0ejmk8ytwrhd3ltm84xehrnlslf
```
Comment (mpoke) [inline app/upgrades/v15/upgrades.go]: Is this supposed to be the same account address as for Discrepancy #1?

File: app/upgrades/v15/upgrades.go [hunk 2]
```diff
@@ -350,3 +356,94 @@ func SetMinInitialDepositRatio(ctx sdk.Context, gk govkeeper.Keeper) error {
...
+func GetEscrowUpdates(ctx sdk.Context) map[string]sdk.Coins {
+	escrowUpdates := map[string]sdk.Coins{
+		// discrepancy #1
+		"cosmos1x54ltnyg88k0ejmk8ytwrhd3ltm84xehrnlslf": {
+			{
+				Denom:  "ibc/4925E6ABA571A44D2BE0286D2D29AF42A294D0FF2BB16490149A1B26EAD33729",
+				Amount: sdk.NewInt(40000000000000000),
+			},
+		},
+		// discrepancy #2
```
Comment (mpoke) [inline app/upgrades/v15/upgrades.go:426]: This matches Discrepancy #3 in the comment above.

File: app/upgrades/v15/upgrades.go [hunk 3]
```diff
@@ -350,3 +356,94 @@ func SetMinInitialDepositRatio(ctx sdk.Context, gk govkeeper.Keeper) error {
...
+	}
+
+	// For discrepancy #3, the missing amount in the escrow account is too large
+	// to be represented using an 64-bit integer. Therefore, it's added to the
+	// escrow updates list under the condition that the amount is successfully
+	// converted to the sdk.Int type.
+	if amt, ok := sdk.NewIntFromString("4388000000000000000000"); !ok {
+		ctx.Logger().Error("can't upgrade missing amount in escrow account: '4388000000000000000000'")
+	} else {
+		escrowUpdates["cosmos1x54ltnyg88k0ejmk8ytwrhd3ltm84xehrnlslf"] = sdk.Coins{
```
Comment (mpoke) [inline app/upgrades/v15/upgrades.go]: Isn't this overwritting the coins added for discrepancy #1?

File: app/upgrades/v15/upgrades_test.go [hunk 4]
```diff
@@ -283,3 +283,33 @@ func TestSetMinInitialDepositRatio(t *testing.T) {
...
 }
+
+func TestUpgradeEscrowAccounts(t *testing.T) {
+	gaiaApp := helpers.Setup(t)
+	ctx := gaiaApp.NewUncachedContext(true, tmproto.Header{})
+
+	bankKeeper := gaiaApp.BankKeeper
+	transferKeeper := gaiaApp.TransferKeeper
+
+	escrowUpdates := v15.GetEscrowUpdates(ctx)
```
Comment (mpoke) [inline app/upgrades/v15/upgrades_test.go:294]: Make sure that all three discrepencies are covered.

**Review by mpoke — COMMENTED** [review]
Please rebase on main and add label to backport to v15.1.x. Also, see my comments below.

File: app/upgrades/v15/upgrades.go (hunk 3 above)
Comment (sainoe) [reply app/upgrades/v15/upgrades.go]: Good catch.

File: app/upgrades/v15/upgrades.go (hunk 1 above)
Comment (sainoe) [reply app/upgrades/v15/upgrades.go]: This escrow account address has two discrepancies for coins of a different denom.

**Review by mpoke — APPROVED** [review]

File: app/upgrades/v15/upgrades.go [hunk 5]
```diff
@@ -350,3 +356,96 @@ func SetMinInitialDepositRatio(ctx sdk.Context, gk govkeeper.Keeper) error {
...
+				Amount: sdk.NewInt(2000),
+			},
+		},
+	}
+
+	// For discrepancy #3, the missing amount in the escrow account is too large
+	// to be represented using an 64-bit integer. Therefore, it's added to the
+	// escrow updates list under the condition that the amount is successfully
+	// converted to the sdk.Int type.
+	if amt, ok := sdk.NewIntFromString("4388000000000000000000"); !ok {
```
Comment (MSalopek) [inline app/upgrades/v15/upgrades.go:439]: Unfortunately there isn't a `sdk.MustNewIntFromString` function so this will do.

**Review by MSalopek — APPROVED** [review]
Found no issues.

LGTM!

**Pull request merged**